### ISSUE TITLE
Allow form elements for updates in update.php

### ISSUE THREAD

**Issue by MrHaroldA** [backdrop/backdrop-issues#1912]
Sometimes you will need user interaction to perform certain updates. For example, Backdrop 1.4.0 has a new installer module, but you need to manually look for it and enable it, which causes confusion to users expecting to have all new features after an update.

This could be handled in different ways, like enabling new features on updates, but "power users" will not be happy about this.

Another way is to ask the user if he wants "Pretty new feature" in the update process, like this:

![image](https://cloud.githubusercontent.com/assets/695291/15423152/81d341f0-1e7c-11e6-8761-ea5e4ddcd17d.png)

The update script is however very basic, so forms won't be validated in any way, but you could use these forms to pass user input to update scripts.

**Comment by MrHaroldA** [comment]
Here's a simple proof of concept, which will require some polish and API documentation, but it works!

https://github.com/backdrop/backdrop/compare/1.x...MrHaroldA:1912/Update-User_input?expand=1

You can test it if you implement this in an install file:

```
/**
 * Optionally enable cool new feature.
 */
function MY_MODULE_update_7000(&$sandbox, $update_form_data) {
  if (isset($update_form_data['enable']) && $update_form_data['enable']) {
    // Enable very cool new feature!
  }
}

function MY_MODULE_update_7000_form() {
  $t = get_t();
  $form = array();

  $form['markup'] = array(
    '#markup' => $t('Enable the cool new feature?')
  );
  $form['enable'] = array(
    '#title' => $t('Yes, please!'),
    '#type' => 'checkbox',
  );

  return $form;
}
```

**Comment by Gormartsen** [comment]
I like an idea!

For extra ideas I would like to see short description, checkbox and URL on well documented feature page.

For example for profiler installer - Link to backdropcms.org where this feature well described with screenshots or even video that explain how to use it.

**Comment by MrHaroldA** [comment]
> short description

You can add as much markup as you'd like; just add it to the MY_MODULE_update_7000_form() "hook".

**Comment by klonos** [comment]
:+1: I like this suggestion too. It would have been great to have in 1.4.0 with the Project Installer which is a perfect use case example.

**Comment by MrHaroldA** [comment]
For this to succeed I'll need a "design" and input on how to feed the data into the update batch process; I'm guessing my solution is a "hack".

@jenlampton, @quicksketch ... thoughts?

**Comment by quicksketch** [comment]
I'm not sure about putting this in update.php. When working with updates, we only have a subset of the total Backdrop API available. Updates should be kept to essentially only SQL queries and config updates, and avoid any module APIs. This is because updates are run for *disabled* modules as well as enabled ones, so you can't use the API of a module safely.

As modules aren't typically loaded during updates, other things such as `hook_modules_enabled()` also won't fire reliably if called from update.php. Although a lot less integrated, it might be better to just direct the user to the modules page to turn on "new" modules, if we could somehow pre-filter down the list to the ones that were just added.

**Comment by klonos** [comment]
How about doing it at the very last step/page of the db updates wizard ("Review log")? By then, (ideally) all updates have finished successfully and the Backdrop API is available again.

The last page already has links for the user to navigate to the front page, `/admin` or the site status report page. So, I was thinking that we can put the "New feature with this version" section there as an additional option to those existing ones.

What do you thing?

**Comment by MrHaroldA** [comment]
How about showing the full update changelog after a Backdrop core update, within your website; with easy links to the sections in your own website to enable stuff like modules, layouts, etc?

**Comment by klonos** [comment]
I would still like us to implement this feature. When introducing new features, such as in #3659, we need a way to allow existing sites to opt in to those. Should there be a separate hook that runs after update.php has finished, and show some sort of message in the site status page or the dashboard?

**Comment by jenlampton** [comment]
I don't know about this...  the database update is something that always needs to be performed, and we shouldn't ever ask the user a question here they aren't prepared answer, or it could cause them not to update, creating massive problems. Everyone's instinct should be to run the update script blindly (because the update script will know better than we do what needs to be done). 

I'd prefer that we find another way to introduce new features that doesn't interfere with the way we perform this super-important task! What if we used the dashboard? That's a much safer place to introduce new features, and it wouldn't disappear once the update was complete.

**Comment by klonos** [comment]
Right. I was talking about something that is triggered _after_ db updates have been performed. Something like an info message on the final screen screen of update.php for example:

![image](https://user-images.githubusercontent.com/2423362/91117610-b20f0380-e697-11ea-9e38-18d425e01f35.png)

...but if we do that, what stops us from making that a form (needs a submit button - this is just a quick mockup):

![image](https://user-images.githubusercontent.com/2423362/91118051-bee02700-e698-11ea-8235-5a2082d27c40.png)

It would be nice if we had #4571 so that we can show one-time, dismissible messages, or perhaps add something in the Dashboard (#3995) and/or the status page.
